src/coreclr/src/jit/hwintrinsiclistarm64.h

```diff
@@ -24,6 +24,8 @@ HARDWARE_INTRINSIC(Vector64,        AsSingle,                                  -
 HARDWARE_INTRINSIC(Vector64,        AsUInt16,                                  -1,               8,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGen|HW_Flag_BaseTypeFromFirstArg)
 HARDWARE_INTRINSIC(Vector64,        AsUInt32,                                  -1,               8,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGen|HW_Flag_BaseTypeFromFirstArg)
 HARDWARE_INTRINSIC(Vector64,        get_Count,                                 -1,               8,           0,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGen)
+HARDWARE_INTRINSIC(Vector64,        get_Zero,                                   0,               8,           0,     {INS_movi,              INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi},              HW_Category_Helper,                 HW_Flag_SpecialImport)
```

Comment:
I believe if you change category to `HW_Category_SimpleSIMD` you wouldn't need the custom importation logic - but you would need `HW_Flag_SpecialCodeGen` (the only flag that is needed here)

Reply:
In the latest commit, I've changed the category to SimpleSIMD and removed the custom importation logic. However, it looks like it is still hitting the `impSpecialCodeGen` method and now because I removed the importation logic the call is not imported as a `GT_HWINTRINSIC` node. Jit dump here:
```
Trees after linear scan register allocator (LSRA)

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1        [000..01D)        (return)                     i label target gcsafe LIR
-----------------------------------------------------------------------------------------------------------------------------------------

------------ BB01 [000..01D) (return), preds={} succs={}
N002 (???,???) [000021] ------------                 IL_OFFSET void   IL offset: 0x0 REG NA
N004 (  1,  1) [000013] ------------        t13 =    CNS_INT(h) long   0x4001BB8 [ICON_STR_HDL] REG x0
                                                  ┌──▌  t13    long
N006 (  4,  3) [000014] n---G-------        t14 = ▌  IND       ref    REG x0
                                                  ┌──▌  t14    ref
N008 (???,???) [000025] ----G-------        t25 = ▌  PUTARG_REG ref    REG x0
                                                  ┌──▌  t25    ref    arg0 in x0
N010 ( 18,  6) [000001] --CXG-------              ▌  CALL      void   TestLibrary.TestFramework.BeginScenario REG NA
N012 (???,???) [000022] ------------                 IL_OFFSET void   IL offset: 0xa REG NA
N014 ( 14,  2) [000002] --CXG-------         t2 =    CALL      simd16 System.Runtime.Intrinsics.Vector128`1[UInt16][System.UInt16].get_AllBitsSet REG d0
                                                  ┌──▌  t2     simd16
N016 ( 18,  5) [000005] DA-XG-------              ▌  STORE_LCL_VAR simd16<System.Runtime.Intrinsics.Vector128`1[UInt16]> V01 loc0          NA REG NA
N018 (???,???) [000023] ------------                 IL_OFFSET void   IL offset: 0x10 REG NA
N020 (  1,  1) [000016] ------------        t16 =    CNS_INT(h) long   0x4001BB8 [ICON_STR_HDL] REG x1
                                                  ┌──▌  t16    long
N022 (  4,  3) [000017] n---G-------        t17 = ▌  IND       ref    REG x1
                                                  ┌──▌  t17    ref
N024 (???,???) [000026] ----G-------        t26 = ▌  PUTARG_REG ref    REG x1
N026 (  3,  2) [000006] ------------         t6 =    LCL_VAR   ref    V00 this          x0 REG x0
                                                  ┌──▌  t6     ref
N028 (???,???) [000027] ------------        t27 = ▌  PUTARG_REG ref    REG x0
N030 (  3,  2) [000007] ------------         t7 =    LCL_VAR   simd16<System.Runtime.Intrinsics.Vector128`1[UInt16]> V01 loc0          d0 REG d0
                                                  ┌──▌  t7     simd16
N032 (???,???) [000028] ------------        t28 = ▌  PUTARG_REG simd16 REG d0
                                                  ┌──▌  t26    ref    arg2 in x1
                                                  ├──▌  t27    ref    this in x0
                                                  ├──▌  t28    simd16 arg1 in d0
N034 ( 24, 13) [000009] --CXG-------              ▌  CALL      void   JIT.HardwareIntrinsics.General.VectorAllBitsSet__AllBitsSetUInt16.ValidateResult REG NA
N036 (???,???) [000024] ------------                 IL_OFFSET void   IL offset: 0x1c REG NA
N038 (  0,  0) [000012] ------------                 RETURN    void   REG NA

-------------------------------------------------------------------------------------------------------------------
```

Reply:
(@echesakovMSFT)

Reply:
@Gnbrkm41 You changed the category but kept SpecialImport flag. What I have suggested is to change the flags column so it has only SpecialCodeGen instead

> if you change category to HW_Category_SimpleSIMD you wouldn't need the custom importation logic - but you would need HW_Flag_SpecialCodeGen

I guess the way I worded this was confusing...

@tannergooding I wonder if in case when SpecialImport flag is set but there is no custom importation logic we should assert instead of expanding an intrinsic as a call?

Reply:
ah, right, sorry, I must have missed that part 😅

Reply:
The category and helpers are different and have slightly different meanings iirc.
We could probably simplify it a bit, but its likely better to do in a separate PR.

Reply:
RE: custom importation logic - I've managed to fix the assertion failures for the expected types, but now having problem with invalid base types, as the API allows any type for the base type (e.g. `Vector64<bool>.Zero`), and the importing logic for table driven intrinsics doesn't handle arbitrary base types.

Do we modify the `impHWIntrinsic` method to handle that, or just write custom import logic? those two APIs (Zero, AllBitsSet) are kind of a special case so it feels right to just make it `SpecialImport`...

Reply:
Actually, I am not sure what is the problem here. Vector64<bool>.Zero works just fine but the 128 one fails during import.

Reply:
Taking a look.

Reply:
`Vector64<T>` isn't failing because none of the methods are marked `[Intrinsic]` as they are on `Vector128<T>`.

`Vector128<T>` is asserting in `HWIntrinsicInfo::lookupIns` because the `type` isn't one of the 10 supported. `impHWIntrinsic` needs something like the following:
```c++
if (!varTypeIsArithmetic(baseType))
{
    return nullptr;
}
```

It should probably be inserted here, right after we finish determining the baseType: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsic.cpp#L645
* Note: I didn't do any extensive analysis to make sure we aren't missing something for the non-table driven intrinsics

Reply:
> `Vector64<T>` isn't failing because none of the methods are marked [Intrinsic] as they are on `Vector128<T>`. 

Aaaaaaaaah. that explains why I was not hitting any of the breakpoints under `impIntrinsic` when importing Vector64 methods. I never checked the disasm for Vector64 methods and they apparently are not properly intrinsified also... 😅

Reply:
> `impHWIntrinsic` needs something like the following:

Do we want to do this, though? for me it seems to suggest that we do need the special importing logics for it since none of the types expect arbitrary base types to be specified...

Reply:
@Gnbrkm41 I personally would be fine with either approach.

I would add the check that Tanner suggested anyway - this will ensure that if the base type does not correspond to a valid element type of a SIMD vector then treat the method being imported as a simple call not as an intrinsic - this what "return nullptr" technically means.

Reply:
The check is technically already there, in this part:
```cpp
if (!isScalar && ((HWIntrinsicInfo::lookupIns(intrinsic, baseType) == INS_invalid) ||
                    ((simdSize != 8) && (simdSize != 16) && (simdSize != 32))))
{
    assert(!"Unexpected HW Intrinsic");
    return nullptr;
}
```
`HWIntrinsicInfo::lookupIns` asserts that `baseType` is one of the supported types, and additionally returns `INS_invalid` if `baseType` is invalid. So we already do have the check (we won't hard crash in release builds during JIT because of it), it's just that we assume that none of the table driven importations can have invalid base type and what we have now causes assert failure in those places.

Reply:
The assert is there to avoid errors later down in the stack. I think it is fine to duplicate it so we can otherwise maintain the assertion.

Reply:
After doing this, the CLR started hitting an assertion in `CLRException::GetThrowable` ("Recursion in CLRException::GetThrowable"); perhaps the changes we introduced here are returning nullptr too early on invalid baseTypes? I think we need to investigate a little bit more if we want to make those table-driven. For now, marking those as `SpecialImport` and re-introducing the custom importation logics.
![image](https://user-images.githubusercontent.com/42944058/80291603-ecb86c00-8789-11ea-8205-04273a69a8da.png)

Reply:
This is interesting. I have started noticing this recently.

Reply:
The change that caused the assertion failure was reverted, so I doubt this PR contains the cause for that exception / is relevant?

I'd be interested to know what exactly is causing that though.

Reply:
Sure, not blaming this PR :), but there is something that we both did and causing this. This is my callstack.

![image](https://user-images.githubusercontent.com/12488060/81242814-f4b2b000-8fc2-11ea-8dba-fcc811a046fa.png)